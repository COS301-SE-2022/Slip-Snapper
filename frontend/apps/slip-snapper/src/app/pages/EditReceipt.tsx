import { IonContent, IonHeader, IonPage, IonTitle, IonToolbar, IonButtons, IonItem, IonButton, IonCard, IonFooter, IonGrid, IonCardHeader, IonCardTitle, IonCol, IonInput, IonLabel, IonRow, IonIcon, IonAlert } from '@ionic/react';
import React, { useState } from 'react';
import '../theme/addEntry.css';
import { NavButtons } from '../components/NavButtons';
import { add } from 'ionicons/icons';
import { updateSlipA } from '../../api/apiCall';

const EditReciept: React.FC = () => {
    
    const slipContents = JSON.parse(localStorage.getItem('editSlip')!);
    const [editRecieptItems, setEditRecieptItems] = useState(slipContents.items);
    const originalItems = slipContents.items

    const [showAlert, setShowAlert] = useState(false);
    const [alertMessage, setAlertMes] = useState("");
    
    return (
        <IonPage>
            <IonHeader>
                <IonToolbar color="primary">
                    <IonTitle>Edit Reciept</IonTitle>
                    <IonButtons slot="end">
                        <NavButtons />
                    </IonButtons>
                </IonToolbar>
            </IonHeader>
            <IonContent fullscreen>

                <IonCard color="primary">
                    <IonCardHeader>
                        <IonCardTitle>Store Name/Location:
                            <IonItem className='addEntry' color="tertiary">
                                <IonInput value={slipContents.location} id={"Store_Name"} contentEditable="true"></IonInput>
                            </IonItem>
                        </IonCardTitle>
                        <IonCardTitle>Date:
                            <IonItem className='addEntry' color="tertiary">
                                <IonInput value={new Date(slipContents.transactionDate).toLocaleDateString('en-GB')} id={"date"} contentEditable="true"></IonInput>
                            </IonItem>
                        </IonCardTitle>

                    </IonCardHeader>
                    <IonCardHeader>
                        <IonCardTitle>Edit Details</IonCardTitle>
                    </IonCardHeader>
                    <IonGrid>
                        <IonRow>
                            <IonCol></IonCol>

                            <IonCol>
                                <IonLabel>Item Name</IonLabel>
                            </IonCol>

                            <IonCol>
                                <IonLabel>Quantity</IonLabel>
                            </IonCol>

                            <IonCol>
                                <IonLabel>Price</IonLabel>
                            </IonCol>

                            <IonCol>
                                <IonLabel>Type</IonLabel>
                            </IonCol>
                            <IonCol>
                            </IonCol>
                        </IonRow>
                    </IonGrid>

                    {editRecieptItems.map((item: any, index: number) => {
                        return (
                            <IonGrid key={index} >
                                <IonRow>
                                    <IonCol className='itemlabels'>
                                        <IonLabel>Item #{index + 1}</IonLabel>
                                    </IonCol>

                                    <IonCol>
                                        <IonItem data-testid={index + "/item"} color="tertiary" className='inputs'>
                                            <IonInput id={index + "/item"} value={item.data.item}></IonInput>
                                        </IonItem>
                                    </IonCol>

                                    <IonCol>
                                        <IonItem color="tertiary" className='inputs'>
                                            <IonInput type='number'
                                                id={index + "/quantity"} value={item.itemQuantity}  ></IonInput>
                                        </IonItem>
                                    </IonCol>

                                    <IonCol>
                                        <IonItem color="tertiary" className='inputs'>
                                            <IonInput
                                                id={index + "/price"} value={item.itemPrice} ></IonInput>
                                        </IonItem>
                                    </IonCol>

                                    <IonCol>
                                        <IonItem color="tertiary" className='inputs'>
                                            <IonInput  id={index + "/type"} value={item.data.itemType} ></IonInput>
                                        </IonItem>
                                    </IonCol>
                                    <IonCol >
                                        <IonButton size='small' fill="outline" color="secondary" onClick={() => removeItem(index)}>Remove</IonButton>
                                        <IonAlert
                                            isOpen={showAlert}
                                            onDidDismiss={() => setShowAlert(false)}
                                            header="Oops..."
                                            message={alertMessage}
                                            buttons={['Ok']}
                                        
                                        />
                                    </IonCol>
                                </IonRow>
                            </IonGrid>
                        )
                    })}
                    <IonItem color="tertiary" className='slipTotal'>
                        <div className='totalHeader'>Total:</div>
                        <IonItem color="tertiary" className='total'>
                            <IonInput id={"total"} value={slipContents.total} ></IonInput>
                        </IonItem>
                    </IonItem>
                    <IonItem color="primary">
                        <IonButton onClick={addItem} slot="start" color="secondary"><IonIcon src={add}></IonIcon></IonButton>
                        <IonButton id='cancelButton' fill="solid" slot="end" color="medium" routerLink={'/receipts'}>Cancel</IonButton>
                        <IonButton onClick={() => { getData(); validateData(); }} fill="solid" slot="end" color="secondary">Submit</IonButton>
                    </IonItem>
                </IonCard>
            </IonContent>
            <IonFooter>
                <IonToolbar color="primary">
                </IonToolbar>
            </IonFooter>
        </IonPage>
    );

    function addItem() {
        getData()
        setEditRecieptItems([...editRecieptItems, { data: { id: editRecieptItems.length+1, item: "", itemType: "" }, itemPrice: 0, itemQuantity: 1 }])
    }
    function removeItem(index: number) {
        getData()
        const data = [...editRecieptItems];
        if (data.length === 1) {
            setAlertMes("A receipt needs to have at least one item.")
            setShowAlert(true);
        }
        else {
            data.splice(index, 1)
            setEditRecieptItems(data)
        }
    }

    function getData() {
        for (let i = 0; i < editRecieptItems.length; i++) {
            const n = document.getElementById(i + "/item")?.getElementsByTagName("input")[0].value
            const q = document.getElementById(i + "/quantity")?.getElementsByTagName("input")[0].value
            const p = document.getElementById(i + "/price")?.getElementsByTagName("input")[0].value
            const t = document.getElementById(i + "/type")?.getElementsByTagName("input")[0].value

            if (n !== undefined) {
                editRecieptItems[i].data.item = n
            } if (q !== undefined) {
                editRecieptItems[i].itemQuantity = Number(q)
            } if (p !== undefined) {
                editRecieptItems[i].itemPrice = p
            } if (t !== undefined) {
                editRecieptItems[i].data.itemType = t
            }
        }
    }

    function validateData() {
        if (document.getElementById("Store_Name")?.getElementsByTagName("input")[0].value === "") {
            setAlertMes("Please enter a Store Name to continue.")
            setShowAlert(true)
            return
        }
        if (document.getElementById("date")?.getElementsByTagName("input")[0].value === "") {
            setAlertMes("Please enter a Date to continue.")
            setShowAlert(true)
            return
        }
        if (document.getElementById("total")?.getElementsByTagName("input")[0].value === "") {
            setAlertMes("Please enter a Total to continue.")
            setShowAlert(true)
            return
        }

        for (let i = 0; i < editRecieptItems.length; i++) {
            if (editRecieptItems[i].data.item === "" ||
                editRecieptItems[i].data.itemType === "" ||
                !Number.isInteger(editRecieptItems[i].itemQuantity) || 
                Math.sign(editRecieptItems[i].itemQuantity) !== 1 ||
                editRecieptItems[i].price === "") {
                setAlertMes("Please complete all fields for item #" + (i + 1) + " to continue.")
                setShowAlert(true)
                return

            }
        }

        const updateItems = editRecieptItems;

        const removeItems: unknown[] = []
        
        for (const rem of originalItems){
            let flag = false
            for(const item of editRecieptItems){
                if(item.id === rem.id){
                    flag = true
                }
            }
            if(!flag){
                removeItems.push(rem.id)
            }
        }

        const insertItems: unknown[] = []
        for (const item of editRecieptItems){
            
            if(item.id === undefined){
                insertItems.push(item)
            }

        }

        const storeName = document.getElementById("Store_Name")?.getElementsByTagName("input")[0].value
        const date = document.getElementById("date")?.getElementsByTagName("input")[0].value
        const temp = document.getElementById("total")?.getElementsByTagName("input")[0].value
        let total
        if (temp !==undefined)
        {
             total = parseFloat(temp)
        }
        const data = {
            text: [date, storeName, "", "", total, slipContents.id]
        };

        let user = JSON.parse(localStorage.getItem('user')!)
        if(user==null){
            user = {id: 24}
        }
        updateSlipA(user.id, data, insertItems, updateItems, removeItems)
        setEditRecieptItems([{ data: { id: editRecieptItems.length+1, item: "", itemType: "" }, itemPrice: 0, itemQuantity: 1 }])
        
        const button = document.getElementById("cancelButton")
        if (button) {
            button.click();
            window.location.reload()
        }

    }
}



export default EditReciept;