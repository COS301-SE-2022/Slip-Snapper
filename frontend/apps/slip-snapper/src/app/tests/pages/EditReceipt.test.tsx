import renderer from 'react-test-renderer';
import EditReceipt from '../../pages/EditReceipt';

const renderTree = (tree: JSX.Element) => renderer.create(tree);
describe('<EditReceipt>', () => {
  it('should render component', () => {
    expect(renderTree(<EditReceipt 
    />).toJSON()).toMatchSnapshot();
  });
  
});