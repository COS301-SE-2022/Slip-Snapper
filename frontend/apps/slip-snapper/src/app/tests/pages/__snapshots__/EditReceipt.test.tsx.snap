// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditReceipt> should render component 1`] = `
<div
  className="ion-page"
>
  <ion-header>
    <ion-toolbar
      color="primary"
    >
      <ion-title>
        Edit Receipt
      </ion-title>
      <ion-buttons
        slot="end"
      >
        <div>
          <ion-button
            onClick={[Function]}
          >
            Home
          </ion-button>
          <ion-button
            onClick={[Function]}
          >
            Reports
          </ion-button>
          <ion-button
            onClick={[Function]}
          >
            Profile
          </ion-button>
          <ion-button
            onClick={[Function]}
          >
            Receipts
          </ion-button>
        </div>
      </ion-buttons>
    </ion-toolbar>
  </ion-header>
  <ion-content
    fullscreen={true}
  >
    <ion-card
      color="primary"
      onClick={[Function]}
    >
      <ion-card-header>
        <div
          color="primary"
        >
          <ion-card-title>
            Location:
            <ion-item
              color="tertiary"
              onClick={[Function]}
            >
              <ion-input
                content-editable="true"
                id="Store_Name"
              />
            </ion-item>
          </ion-card-title>
        </div>
        <div
          color="primary"
        >
          <ion-card-title>
            Date:
            <ion-item
              color="tertiary"
              onClick={[Function]}
            >
              <ion-datetime
                display-format="DD/MM/YYYY"
                id="date"
              />
              <ion-icon
                icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><title>Calendar</title><rect stroke-linejoin='round' x='48' y='80' width='416' height='384' rx='48' class='ionicon-fill-none ionicon-stroke-width'/><circle cx='296' cy='232' r='24'/><circle cx='376' cy='232' r='24'/><circle cx='296' cy='312' r='24'/><circle cx='376' cy='312' r='24'/><circle cx='136' cy='312' r='24'/><circle cx='216' cy='312' r='24'/><circle cx='136' cy='392' r='24'/><circle cx='216' cy='392' r='24'/><circle cx='296' cy='392' r='24'/><path stroke-linejoin='round' stroke-linecap='round' d='M128 48v32M384 48v32' class='ionicon-fill-none ionicon-stroke-width'/><path stroke-linejoin='round' d='M464 160H48' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
                slot="end"
              />
            </ion-item>
          </ion-card-title>
        </div>
      </ion-card-header>
      <ion-card-header>
        <ion-card-title>
          Edit Details
        </ion-card-title>
      </ion-card-header>
      <div
        className="wrapper"
      >
        <div
          className="MuiButtonBase-root MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorSecondary MuiChip-clickable MuiChip-clickableColorSecondary MuiChip-filledSecondary css-1nukl5-MuiButtonBase-root-MuiChip-root"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconMedium MuiChip-iconColorSecondary css-i4bv87-MuiSvgIcon-root"
            data-testid="AddCircleOutlineIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
          <span
            className="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
          >
            New Item
          </span>
        </div>
      </div>
      <ion-card-header>
        <ion-card-title>
          Total Amount:
        </ion-card-title>
      </ion-card-header>
      <ion-card-header>
        <ion-item
          color="tertiary"
          id="total"
          onClick={[Function]}
        />
      </ion-card-header>
      <ion-item
        color="primary"
        onClick={[Function]}
      >
        <ion-button
          color="medium"
          fill="solid"
          id="cancelButton"
          onClick={[Function]}
          slot="end"
        >
          Cancel
        </ion-button>
        <ion-button
          color="secondary"
          fill="solid"
          onClick={[Function]}
          slot="end"
        >
          Submit
        </ion-button>
      </ion-item>
    </ion-card>
  </ion-content>
  <ion-footer>
    <ion-toolbar
      color="primary"
    />
  </ion-footer>
</div>
`;
